File: projects/engineering.purdue.edu/scriptingwithobjects/swocode/chap14/ThreadCooperationWithSleep.py
#!/usr/bin/env python

###  ThreadCooperationWithSleep.py

import threading                                                     #(A)
import time                                                          #(B)
import random                                                        #(C)
ran = random.Random()                                                #(D)

class DoitThread( threading.Thread ):                                #(E)

    def __init__( self, name = id ):                                 #(F)
        threading.Thread.__init__(self)                              #(G)

    def run( self ):                                                 #(H)
        self._do_it()                                                #(I)

    def keepBusy( self, howmany_secs ):                              #(J)
        curr = int( time.time() )                                    #(K)
        while ( int(time.time()) < curr + howmany_secs ): pass       #(L)

    def _do_it(self):                                                #(M)
        threadName = self.getName()                 # Thread-N       #(N)
        hyphenIndex = threadName.index( '-' )                        #(O)
        threadID = threadName[hyphenIndex+1:]                        #(P)
        self.keepBusy( ran.randint( 0, 4 ) )                         #(Q)
        # Threads with even ID's to sleep for 10 seconds:
        if int(threadID) % 2 == 0: time.sleep( 10 )                  #(R)
        # Get just the last two digits of the 
        # current time in seconds:
        curr = int( time.time() ) % 60                               #(S)
        print "do_it executed by thread %s at time %d" % \
                                          (threadID, curr)           #(T)

threads = [ DoitThread(i) for i in range(1,11) ]                     #(U)
map( lambda x: x.start(), threads )                                  #(V)

File: projects/demos/engineering.purdue.edu/scriptingwithobjects/swocode/chap14/ThreadCooperationWithSleep.pl
#!/usr/bin/perl -w

###  ThreadCooperationWithSleep.pl

use strict;
use threads;                                                         #(A)

my @threads;                                                         #(B)
foreach my $i (1 .. 10) {                                            #(C)
    $threads[$i] = threads->new( \&do_it );                          #(D)
}

foreach my $i (1 .. 10) {                                            #(E)
    $threads[$i]->join;                                              #(F)
}

sub keepBusy {                                                       #(G)
    my $how_many_seconds = shift;                                    #(H)
    my $curr = time;                                                 #(I)
    while ( time < $curr + $how_many_seconds ) {}                    #(J)
}

sub do_it {                                                          #(K)
    my $threadID = threads->self->tid();                             #(L)
    keepBusy( int(rand(5)) );                                        #(M)
    # Threads with even ID's to sleep for 10 seconds:
    sleep( 10 ) unless $threadID % 2;                                #(N)
    my $curr = time;                                                 #(O)
    # Get just the last two digits of the 
    # current time in seconds:
    my ($seconds) = $curr =~ /(\d\d)$/;                              #(P)
    print "do_it executed by thread $threadID at time $seconds\n";   #(Q)
}
